proof_of_concept/definitions/asset_id.py

```diff
@@ -0,0 +1,82 @@
+"""Identifier for assets."""
+from typing import Any, cast, Type
+
+
+class AssetId(str):
+    """An ID of an asset."""
+    def __new__(cls: Type['AssetId'], seq: Any) -> 'AssetId':
+        """Create an AssetId.
+
+        Args:
+            seq: Contents, will be converted to a string using str(),
+            then used as the id.
+
+        Raises:
+            ValueError: If str(seq) is not a valid ID.
+        """
+        data = str(seq)
+        if data.startswith('id:'):
+            parts = len(data.split(':'))
+            if parts < 3 or parts > 4:
+                raise ValueError(f'Invalid asset id: {data}')
+        elif data.startswith('hash:'):
+            if not all([c in '0123456789abcdef' for c in data[5:]]):
+                raise ValueError(f'Invalid asset id: {data}')
+        elif data == '*':
+            # Valid as a wildcard in rules
+            pass
+        else:
+            raise ValueError(f'Invalid asset id type: {data}')
+
+        return str.__new__(cls, seq)        # type: ignore
+
+    @staticmethod
+    def from_key(key: str) -> 'AssetId':
+        """Creates an AssetId from a key (hash).
+
+        Args:
+            key: A hash of a workflow that created this asset.
+
+        Returns:
+            The AssetId for the workflow result.
+        """
+        return AssetId(f'hash:{key}')
+
+    def is_primary(self) -> bool:
+        """Returns whether this is a primary asset."""
+        return self.startswith('id:')
+
+    def is_concrete(self) -> bool:
+        """Returns whether this is a concrete asset.
+
+        An asset is concrete if it is primary and has a site it can
+        be downloaded from.
+        """
+        return self.is_primary() and len(self.split(':')) == 4
```

Comment:
Could we think of a better name than `is_concrete`? `has_location` maybe? Something that tells us that it has a location.

Reply:
Well, secondary assets (intermediate and final workflow results) also have a location, but not in their id. It's a subdivision of primary assets, which can be actual data sets and software (concrete) or collections of such (abstract). So maybe `has_location` is also misleading. Hmm.

Reply:
Fair enough. If we cannot think of something better than `is_concrete` is the best name!